### PR TITLE
DrivAerML: stochastic input feature dropout (p=0.05, p=0.10)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Independent dropout on node input features (geometry coordinates, normals, etc.) during training prevents the model from relying on any single geometric feature descriptor, forcing it to learn redundant representations and improving generalization. This is analogous to DropConnect applied at the input layer. We test two dropout rates: p=0.05 (light) and p=0.10 (moderate). Dropout is only active during training (`self.training=True`), never at eval time.

The intuition: if the model learns to predict pressure even when some geometric features are randomly masked, it must build a more robust internal representation. This is a common regularization technique in tabular/feature-based ML (used widely in Kaggle competitions) and has been shown to help in geometric deep learning when input features are correlated.

## Instructions

Add an `--input-feature-dropout` float argument to `train.py` (default `0.0`). Apply `F.dropout(x, p=args.input_feature_dropout, training=model.training)` to the node input feature tensor `x` immediately **before** it is passed into the first attention/message-passing layer. Do **not** apply this to edge features or any intermediate hidden representations — only the raw node input features.

Run two variants back-to-back in the same job using `--wandb_group dm-input-dropout`:

**Variant A — p=0.05 (light dropout):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --input-feature-dropout 0.05 \
  --wandb_group dm-input-dropout --wandb_name dm-input-dropout-p005
```

**Variant B — p=0.10 (moderate dropout):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --input-feature-dropout 0.10 \
  --wandb_group dm-input-dropout --wandb_name dm-input-dropout-p010
```

Run both variants to at least 300 epochs (or until the cosine schedule completes at T_max=30 cycles), whichever comes first. Report `val_primary/surface_rel_l2_pct` for both variants at the best validation checkpoint.

## Baseline

Current best DrivAerML metrics (PR #2898, piccolo — 4L/512d/8H + Fourier + no-EMA + T_max=30):
- `val_primary/surface_rel_l2_pct = 3.997%` at epoch 467
- Baseline W&B run: `bht6h42t`
- External target: <3.71% (AB-UPT, ~500 epochs) — 1.08x gap remaining

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Summary

| Variant | p | Best Val (%) | Best Epoch | Best Test (%) | W&B Run | Diverged? |
|---------|---|-------------|-----------|--------------|---------|-----------|
| Baseline | 0.0 | **3.997** | 467 | — | `bht6h42t` | No |
| A | 0.05 | 7.269 | 190 | 7.982 | `75zw9pyt` | Yes (epoch ~320) |
| B | 0.10 | 11.723 | 112 | 12.125 | `v0hdvf5o` | Yes (epoch ~150) |

**Neither variant beats the baseline.** Both variants also caused training instability and eventual divergence.

### Variant A — p=0.05 (light dropout)

- **Best val:** `val_primary/surface_rel_l2_pct = 7.269%` at epoch 190
- **Best test:** `test_primary/surface_rel_l2_pct = 7.982%` at epoch 190
- **W&B run:** [`75zw9pyt`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/75zw9pyt)
- Training was stable through ~epoch 210. After that, grad norms began climbing:
  - Epoch 240: grad_norm=4.73, val=11.4%
  - Epoch 300: grad_norm=6.99, val=22.3%
  - Epoch 330: grad_norm=46.8, val=43.7%
  - Epoch 390: grad_norm=4129, val=49.1%
  - Epoch 506 (final): grad_norm=63B, val=48.6%
- The best metric at epoch 190 was already 1.82x worse than baseline (7.27% vs 3.997%).

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --input-feature-dropout 0.05 \
  --agent fern --wandb-group dm-input-dropout --wandb-name "fern/dm-input-dropout-p005"
```

### Variant B — p=0.10 (moderate dropout)

- **Best val:** `val_primary/surface_rel_l2_pct = 11.723%` at epoch 112
- **Best test:** `test_primary/surface_rel_l2_pct = 12.125%` at epoch 112
- **W&B run:** [`v0hdvf5o`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/v0hdvf5o)
- Diverged much earlier and more severely:
  - Epoch 150: grad_norm=3.48, val=38.9%
  - Epoch 180: grad_norm=95, val=47.7%
  - Epoch 300: grad_norm=1643, val=37.5%
  - Epoch 398: loss=NaN (completely diverged)
- Best metric at epoch 112 was 2.93x worse than baseline (11.72% vs 3.997%).

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --input-feature-dropout 0.10 \
  --agent fern --wandb-group dm-input-dropout --wandb-name "fern/dm-input-dropout-p010"
```

### What happened

**Input feature dropout clearly hurts DrivAerML performance and causes training instability.** Key observations:

1. **No improvement at any point during training.** Even at their best epochs (before divergence), neither variant approached the baseline. The best p=0.05 achieved was 7.27% vs baseline 3.997%.

2. **Dose-dependent instability.** Higher dropout (p=0.10) diverged earlier (epoch ~150 vs ~320) and achieved worse best metrics (11.72% vs 7.27%). This is a clean negative dose-response.

3. **Divergence mechanism.** The dropout noise on geometric input features (coordinates, normals) appears to destabilize the gradient signal over long training. Initially training progresses normally, but the accumulated perturbation to the input manifold causes grad norms to grow exponentially, eventually leading to NaN (Variant B) or extreme instability (Variant A).

4. **Why this technique doesn't transfer from tabular ML.** In tabular settings, features are often redundant and weakly correlated. In CFD meshes, geometric features (coordinates, normals, signed distances) are strongly interdependent and physically meaningful — randomly zeroing them creates physically impossible input configurations that the model cannot learn from robustly.

### Bug fix included

Fixed a pre-existing `UnboundLocalError` bug where `primary_metric_key` (a function) was shadowed by a local variable assignment at lines 1646-1649 of train.py. The duplicate assignment block was removed and the already-computed `best_val_primary_metric_name` is now used at line 1723.

### Suggested follow-ups

- **If input regularization is still of interest**, consider alternatives that respect the geometric structure: (a) coordinate jitter (small Gaussian noise on xyz, preserving feature structure), (b) feature-wise Gaussian noise instead of dropout (scaling rather than zeroing), or (c) manifold-aware augmentation that perturbs the mesh geometry consistently rather than dropping individual features.
- **Gradient clipping** might have prevented the divergence but would not have addressed the fundamental accuracy gap.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3149 because ADVISOR: Feature dropout dead — zeroing physically interdependent geometric features creates impossible inputs. 7.269%/11.723% both catastrophically worse. Bug fix noted. Reassigning to noam-branch features on DM champion..
